### PR TITLE
On 971 show manual certificates in previous history tab and staff app

### PR DESCRIPTION
**Pull request by jcoxhead**

### CODE REVIEW

File: src/SFA.DAS.AssessorService.Application/Handlers/Certificates/GetPrivateCertificatePrivateByUlnHandler.cs [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+        public GetPrivateCertificatePrivateByUlnHandler(ICertificateRepository certificateRepository)
+        {
+            _certificateRepository = certificateRepository;
+        }
+
+        public async Task<Certificate> Handle(GetPrivateCertificatePrivateByUlnRequest request,
+            CancellationToken cancellationToken)
+        {
+            var certificate = await _certificateRepository.GetPrivateCertificate(request.Uln, request.EndPointAssessorOrganisationId, request.LastName);
+//            var certificateResponse = Mapper.Map<Certificate, CertificateResponse>(certificate);
```
Comment (MarkFCain) [inline src/SFA.DAS.AssessorService.Application/Handlers/Certificates/GetPrivateCertificatePrivateByUlnHandler.cs]: unneeded comment

File: src/SFA.DAS.AssessorService.Database/Tables/Certificates.sql [hunk 2]
```diff
@@ -17,6 +17,7 @@
     [ProviderUkPrn] INT NOT NULL, 
     [CertificateReferenceId] INT NOT NULL IDENTITY(10001,1), 
 	[LearnRefNumber] NVARCHAR(12) NULL,
+	[ReasonForRejection] [nvarchar](max) NOT NULL,
```
Comment (MarkFCain) [inline src/SFA.DAS.AssessorService.Database/Tables/Certificates.sql]: check you need this

File: src/SFA.DAS.AssessorService.Application/Handlers/Staff/LearnerDetailHandler.cs [hunk 3]
```diff
@@ -52,12 +64,12 @@ public async Task<LearnerDetail> Handle(LearnerDetailRequest request, Cancellati
...
                 }
-                
+
 
                 certificateData = JsonConvert.DeserializeObject<CertificateData>(certificate.CertificateData);
                 epao = await _organisationRepository.Get(certificate.OrganisationId);
             }
-            
+
             var learnerDetail = new LearnerDetail()
```
Comment (MarkFCain) [inline src/SFA.DAS.AssessorService.Application/Handlers/Staff/LearnerDetailHandler.cs]: Just a suggestion.  I would put this into a private mapper method to eliminate the bloat inside this function.  Your choice though.

File: src/SFA.DAS.AssessorService.Database/Tables/Certificates.sql (hunk 2 above)
Comment (jcoxhead) [reply src/SFA.DAS.AssessorService.Database/Tables/Certificates.sql]: agreed tx will fix now

File: src/SFA.DAS.AssessorService.Application/Handlers/Staff/LearnerDetailHandler.cs (hunk 3 above)
Comment (jcoxhead) [reply src/SFA.DAS.AssessorService.Application/Handlers/Staff/LearnerDetailHandler.cs]: yep hang over from past from daves code will see if I get time to resolve for tommorrow.

File: src/SFA.DAS.AssessorService.Api.Types/Models/Certificates/GetApprovedDraftCertificatesRequest.cs [hunk 4]
```diff
@@ -0,0 +1,10 @@
+﻿using System.Collections.Generic;
+using MediatR;
+
+namespace SFA.DAS.AssessorService.Api.Types.Models.Certificates
+{
+    public class GetApprovedDraftCertificatesRequest : IRequest<List<CertificateSummaryResponse>>
+    {
```
Comment (martin-hornsby) [inline src/SFA.DAS.AssessorService.Api.Types/Models/Certificates/GetApprovedDraftCertificatesRequest.cs]: Is this empty by design? If so can you pop a quick comment in here.

File: src/SFA.DAS.AssessorService.Application.Api/Controllers/CertificateController.cs [hunk 5]
```diff
@@ -107,5 +107,15 @@ public async Task<IActionResult> Approvals([FromBody] CertificateApprovalRequest
...
             return Ok();
         }
+
+        [HttpPost("prepareprivatecertificatesforbatchrun", Name = "PreparePrivateCertificatesforBatchRun")]
+        [SwaggerResponse((int)HttpStatusCode.OK, Type = typeof(Certificate))]
+        [SwaggerResponse((int)HttpStatusCode.BadRequest, typeof(IDictionary<string, string>))]
+        [SwaggerResponse((int)HttpStatusCode.InternalServerError, Type = typeof(ApiResponse))]
+        public async Task<IActionResult> PreparePrivateCertificatesforBatchRun([FromBody] PrivateCertificatePrepareForBatchRunRequest privateCertificatePrepareForBatchRunRequest)
+        {
+            await _mediator.Send(privateCertificatePrepareForBatchRunRequest);
```
Comment (martin-hornsby) [inline src/SFA.DAS.AssessorService.Application.Api/Controllers/CertificateController.cs]: The await is outside of the return OK() - is this by design?

File: src/SFA.DAS.AssessorService.Application/Handlers/Certificates/PrivateCertificatePrepareForBatchRunHandler.cs [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+            var certificates = await _certificateRepository.GetCertificates(statuses);
+            foreach (var certificate in certificates)
+            {
+                var certificateData = JsonConvert.DeserializeObject<CertificateData>(certificate.CertificateData);
+                if (certificate.Status == Domain.Consts.CertificateStatus.Approved)
+                {
+                    certificate.Status = Domain.Consts.CertificateStatus.Submitted;
+                }
+
+                if (certificate.Status == Domain.Consts.CertificateStatus.Rejected)
```
Comment (martin-hornsby) [inline src/SFA.DAS.AssessorService.Application/Handlers/Certificates/PrivateCertificatePrepareForBatchRunHandler.cs]: Whilst not a issue right now. Think this should be an else if block

File: src/SFA.DAS.AssessorService.Api.Types/Models/Certificates/GetApprovedDraftCertificatesRequest.cs (hunk 4 above)
Comment (jcoxhead) [reply src/SFA.DAS.AssessorService.Api.Types/Models/Certificates/GetApprovedDraftCertificatesRequest.cs]: done

File: src/SFA.DAS.AssessorService.Application.Api/Controllers/CertificateController.cs (hunk 5 above)
Comment (jcoxhead) [reply src/SFA.DAS.AssessorService.Application.Api/Controllers/CertificateController.cs]: yep there's no response with it.

File: src/SFA.DAS.AssessorService.Application/Handlers/Certificates/PrivateCertificatePrepareForBatchRunHandler.cs (hunk 6 above)
Comment (jcoxhead) [reply src/SFA.DAS.AssessorService.Application/Handlers/Certificates/PrivateCertificatePrepareForBatchRunHandler.cs]: in this case I would have to leave a default case which is more code, so have not done this.

**Comment by alandburns** [comment]
Reworked elsewhere

**Pull request closed without merging**